Bare metal example with memory optimisation enhancements (#27)
ARMmbed/mbed-os-example-blinky-baremetal: README.md

1. Run the following command to build the example project and program the microcontroller flash memory:

   ```bash
   $ mbed compile -m <TARGET> -t <TOOLCHAIN> --flash --sterm
   ```
   
Note: You can use the Mbed CLI command-line option "--sterm" to open a serial terminal after flashing.

Your PC may take a few minutes to compile your code.

The binary is located at `./BUILD/<TARGET>/<TOOLCHAIN>/mbed-os-example-blinky-baremetal.bin`.

Alternatively, you can manually copy the binary to the target, which gets mounted on the host computer through USB.

## Expected output 

The LED on your target turns on and off every 500 milliseconds and the serial terminal shows an output similar to: 

``` 
--- Terminal on /dev/tty.usbmodem21102 - 9600,8,N,1 ---
This is the bare metal blinky Mbed OS example.
``` 

## Configuring the application

### The bare metal mode

The bare metal mode is a configuration of Mbed OS that excludes the RTOS as well as other features. It is specifically targeted for memory constrained devices because it gives developers higher degree of control of their system. See more details [here](https://os.mbed.com/docs/mbed-os/v5.15/reference/mbed-os-bare-metal.html)

> To build the bare metal profile, `mbed_app.json` is specified as follows:

Reviewer: `
```suggestion
To build with the bare metal profile, the configuration file must contain the following:
```
Perhaps a link to information on configuration files is needed here.
The existing text may suggest it has to be named `mbed_app.json`.

```
{
    "requires": ["bare-metal"]
}
```

### Futher optimisations

#### Linking with smaller C libraries

> Both the `ARM` and `GCC_ARM` toolchains support code optimised versions of their C standard libraries, microlib and newlib-nano. It is safe to use these smaller C libraries in bare metal mode and we recommend using them.

Reviewer: ```suggestion
Both the `ARM` and `GCC_ARM` toolchains support optimised versions of the C standard library, [microlib](link) and [newlib-nano](link) respectively, we recommend using them with the bare metal mode.
```

Author: I will take the text recommendation but not provide the links as they would need to be updated with new versions of the compilers.

> To build with the smaller C libraries, `mbed_app.json` needs to be modified as follows:

Reviewer: ```suggestion
To build with the smaller C libraries, the configuration file needs to be modified as follows:
```

```
{
    "target_overrides": {
        "*": {
            "target.c_lib": "small"
        }
    }
}
```
#### Using Mbed minimal printf library

> A smaller printf() alternative is supported in Mbed. The [minimal printf](https://github.com/ARMmbed/mbed-os/blob/master/platform/source/minimal-printf/README.md) library implements a subset of the printf() features and floating points can be disabled to save additional flash.

Reviewer: ```suggestion
Mbed OS supports the [minimal printf](https://github.com/ARMmbed/mbed-os/blob/master/platform/source/minimal-printf/README.md) library, a configurable implementation of a subset of the `v/s/f/printf` function family, to save additional code size.
```

> To build with the minimal printf library and disable floating points printing, `mbed_app.json` needs to be modified as follows:

Reviewer: ```suggestion
To build with the minimal printf library and disable floating points printing, the configuration file needs to be modified as follows:
```

```
{
    "target_overrides": {
        "*": {
            "target.printf_lib": "minimal-printf",
            "platform.minimal-printf-enable-floating-point": "false"
        }
    }
}
```

#### Using a minimal console

> If your application only needs basic unbuffered I/O operations then you can save additional memory by using a configuration of the platform library which removes file handling functionality from the retarget code.

Reviewer: ```suggestion
If your application only needs basic unbuffered I/O operations then you can save additional memory by using a configuration of the platform library which removes file handling functionality from the [system I/O retarget code](/path/to/mbed_retarget.cpp).
```

> To build with the minimal console functionality, `mbed_app.json` needs to be modified as follows:

Reviewer: ```suggestion
To build with the minimal console functionality, the configuration file needs to be modified as follows:
```

```
{
    "target_overrides": {
        "*": {
            "platform.stdio-minimal-console-only": "true"
        }
    }
}
```

#### Memory comparison

> Example configuration files are provide under the `configs/` directory. You may override the `mbed_app.json` with either of these.

Reviewer: ```suggestion
Example configuration files are provide under the `configs/` directory. You may override the `mbed_app.json` with either of these using the `--appconfig` optional argument.
```

> The below table shows result for the blinky bare metal application compiled with the release profile on K64F for two toolchains.

Reviewer: Rather than showing the table and sizes, it was previously recommended to instead mention the number of bytes saving.

   ```
   
Note: You can use the Mbed CLI command-line option "--sterm" to open a serial terminal after flashing.

Your PC may take a few minutes to compile your code.

The binary is located at `./BUILD/<TARGET>/<TOOLCHAIN>/mbed-os-example-blinky-baremetal.bin`.

Alternatively, you can manually copy the binary to the target, which gets mounted on the host computer through USB.

## Expected output 

The LED on your target turns on and off every 500 milliseconds and the serial terminal shows an output similar to: 

``` 
--- Terminal on /dev/tty.usbmodem21102 - 9600,8,N,1 ---
This is the bare metal blinky Mbed OS example.
``` 

## Configuring the application

### The bare metal mode

The bare metal mode is a configuration of Mbed OS that excludes the RTOS as well as other features. It is specifically targeted for memory constrained devices because it gives developers higher degree of control of their system. See more details [here](https://os.mbed.com/docs/mbed-os/v5.15/reference/mbed-os-bare-metal.html)

To build with the bare metal profile, the application configuration file must contain the following:
```
{
    "requires": ["bare-metal"]
}
```

> Note: the `mbed_app.json` may be present at the root of your application or specified as the argument of the --app-config parameter. For convenience, a few configuration files are present in the `configs` directory so you can try some of the different configurations shown in this example.

Reviewer: ```suggestion
Note: the application configuration is picked up by the build tool if named `mbed_app.json`. Alternatively, it can be given a more descriptive name and provided to the build tool using an optional argument to the `compile` sub-command. See [here](link to Memory comparison section) for how to provide a different application configuration file to the build tool.
```
Something like that.
Reasons for change request:
1. `mbed_app.json` does not have to be in the root, it gets picked up anywhere if it is called `mbed_app.json` (https://os.mbed.com/docs/mbed-os/v5.15/tools/greentea-testing-applications.html see section `Application configuration`)
2. `--app-config` is mentioned further down, I suggest we remove it from here to avoid repetition.

Author: The sentence was directly taken from our documentation https://os.mbed.com/docs/mbed-os/v5.15/reference/configuration.html. I don't want this to become a discussion on configuration. I will just remove the first sentence to remove the repetition on `--app-config`

![](./resources/official_armmbed_example_badge.png)
# Bare metal blinky Mbed OS example

This example shows how to achieve memory optimisations in Mbed. Starting with a simple blinky application, the example illustrates how to enable the bare metal profile and further memory optimisations.

You can build this project with all supported [Mbed OS build tools](https://os.mbed.com/docs/mbed-os/latest/tools/index.html). However, this example project specifically refers to the command-line interface tool [Arm Mbed CLI](https://github.com/ARMmbed/mbed-cli#installing-mbed-cli).

## Application functionality

The `main()` function toggles the state of a digital output connected to an LED on the board.

## Prerequisites

> 

Reviewer 2: Can we keep the generic parts of the README the same please https://github.com/ARMmbed/mbed-os-example-blinky/blob/master/README.md

To build with the bare metal profile, the application configuration file must contain the following:
```
{
    "requires": ["bare-metal"]
}
```

Note: For convenience, a few configuration files are present in the `configs` directory so you can try some of the different configurations shown in this example.

### Futher optimisations

#### Linking with smaller C libraries

Both the `ARM` and `GCC_ARM` toolchains support optimised versions of their C standard libraries, microlib and newlib-nano respectively and we recommend using them with the bare metal profile.

To build with the smaller C libraries, the application configuration file needs to be modified as follows:
```
{
    "target_overrides": {
        "*": {
            "target.c_lib": "small"
        }
    }
}
```
#### Using Mbed minimal printf library

Mbed OS offers a smaller printf() alternative. The [minimal printf](https://github.com/ARMmbed/mbed-os/blob/master/platform/source/minimal-printf/README.md) library implements a subset of the `v/s/f/printf` function family and floating points can be disabled to save additional code size.

> To build with the minimal printf library and disable floating points printing, the application configuration file needs to be modified as follows:

Reviewer 2: We can add a note that further optimisations are possible and details are in README.

```
{
    "target_overrides": {
        "*": {
            "target.printf_lib": "minimal-printf",
            "platform.minimal-printf-enable-floating-point": "false"
        }
    }
}
```

#### Using a minimal console

If your application only needs basic unbuffered I/O operations then you can save additional memory by using a configuration of the platform library which removes file handling functionality from the [system I/O retarget code](https://github.com/ARMmbed/mbed-os/blob/master/platform/source/mbed_retarget.cpp).

To build with the minimal console functionality, the application configuration file needs to be modified as follows:
```
{
    "target_overrides": {
        "*": {
            "platform.stdio-minimal-console-only": "true"
        }
    }
}
```

#### Memory comparison

> Example configuration files are provided under the `configs/` directory. You may override the `mbed_app.json` with either of these using the `--appconfig` optional argument.

Reviewer 2: Is that useful (providing additional config for each option)? Or is it just noise and complication?

Author: I can remove indeed. I was in two minds about it.
